File: test/features/manualentry/activity/recorded_session_range_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:openvitals/domain/model/ble_sensor_models.dart';
import 'package:openvitals/domain/preferences/unit_system.dart';
import 'package:openvitals/features/manualentry/activity/activity_entry_clock.dart';
import 'package:openvitals/features/manualentry/activity/activity_entry_state.dart';
import 'package:openvitals/features/manualentry/activity/activity_entry_write_request_builder.dart';

/// The session written to Health Connect has to CONTAIN the samples it carries.
///
/// A recording reaches the entry form as TEXT, at minute granularity: the start time
/// loses its seconds, and the duration is rounded up to a whole minute. The write path
/// then rebuilds the session range from that text — so the rebuilt end can land before
/// the last sample that was actually recorded.
///
/// Health Connect does not drop a sample that falls outside its session. It CLAMPS it to
/// the session bounds. So the samples past the end are not lost, they are stacked onto
/// the closing instant, all sharing one timestamp — which is worse than losing them, and
/// is invisible unless you go looking. For a bike ride it is a strange-looking tail. For
/// a heart-rate recovery, computed from precisely those last samples, it is everything.

BleRecordingSampleBuffer _heartRate(List<(DateTime, int)> beats) =>
    BleRecordingSampleBuffer(
      heartRateSamples: [
        for (final (time, bpm) in beats)
          BleHeartRateSample(time: time, beatsPerMinute: bpm),
      ],
    );

ActivityEntryUiState _recorded({
  required String startTimeText,
  required String durationMinutesText,
  required BleRecordingSampleBuffer samples,
}) {
  final clock = ActivityEntryClock.system();
  return initialActivityEntryState(clock, const {'write-exercise'}).copyWith(
    startDateText: '2026-07-14',
    startTimeText: startTimeText,
    durationMinutesText: durationMinutesText,
    isRecordingDraft: true,
    recordedBleSamples: samples,
  );
}

void main() {
  test('the session is stretched to cover the last recorded sample', () {
    // The worked example: recording started at 10:00:59 and ran 120 seconds. The form
    // truncates the start to 10:00:00 and rounds the duration to 2 minutes, so the
    // rebuilt session ends at 10:02:00 — while the last heart-rate sample was taken at
    // 10:02:59, a full 59 seconds later.
    final request = buildWriteRequest(
      _recorded(
        startTimeText: '10:00',
        durationMinutesText: '2',
        samples: _heartRate([
          (DateTime(2026, 7, 14, 10, 0, 59), 150),
          (DateTime(2026, 7, 14, 10, 1, 59), 165),
          (DateTime(2026, 7, 14, 10, 2, 59), 172),
        ]),
      ),
      UnitSystem.metric,
    );

    expect(request, isNotNull);
    final lastSample = DateTime(2026, 7, 14, 10, 2, 59);
    expect(
      request!.endTime.isAfter(lastSample),
      isTrue,
      reason: 'the session must end after its own last sample, or Health Connect '
          'collapses that final minute of readings onto a single instant',
    );
    // And the samples themselves are still all there.
    expect(request.bleSamples.heartRateSamples, hasLength(3));
  });

  test('an untruncated recording is left exactly as it is', () {
    final request = buildWriteRequest(
      _recorded(
        startTimeText: '10:00',
        durationMinutesText: '30',
        samples: _heartRate([
          (DateTime(2026, 7, 14, 10, 5), 150),
          (DateTime(2026, 7, 14, 10, 25), 160),
        ]),
      ),
      UnitSystem.metric,
    );

    expect(request!.endTime, DateTime(2026, 7, 14, 10, 30),
        reason: 'nothing to stretch: the samples already fit');
    expect(request.bleSamples.heartRateSamples, hasLength(2));
  });

  test('samples before a start the user moved forward are dropped, not clamped', () {
    // The user recorded from 10:00, then edited the start to 10:10 before saving. The
    // early samples cannot be kept: Health Connect would clamp them ONTO 10:10, inventing
    // a burst of readings that were never taken at that time.
    final request = buildWriteRequest(
      _recorded(
        startTimeText: '10:10',
        durationMinutesText: '20',
        samples: _heartRate([
          (DateTime(2026, 7, 14, 10, 0), 120),
          (DateTime(2026, 7, 14, 10, 15), 150),
        ]),
      ),
      UnitSystem.metric,
    );

    expect(request, isNotNull, reason: 'the workout still saves');
    expect(
      request!.bleSamples.isEmpty(),
      isTrue,
      reason: 'better to write no heart rate than to write a reading at a time it '
          'was never taken',
    );
  });
}

File: app/src/test/kotlin/tech/mmarca/openvitals/features/manualentry/activity/RecordedSessionRangeTest.kt
package tech.mmarca.openvitals.features.manualentry.activity

import java.time.Instant
import java.time.LocalDateTime
import java.time.ZoneId
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import tech.mmarca.openvitals.domain.model.BleHeartRateSample
import tech.mmarca.openvitals.domain.model.BleRecordingSampleBuffer
import tech.mmarca.openvitals.domain.preferences.UnitSystem

/**
 * The session written to Health Connect has to contain the samples it carries. The form
 * rebuilds the range from minute-granularity text, so the end can land before the last
 * sample. Health Connect clamps such samples onto the closing instant, which ruins a recovery.
 */
class RecordedSessionRangeTest {

    private val zone = ZoneId.systemDefault()
    private val start: Instant =
        LocalDateTime.of(2026, 7, 14, 10, 0).atZone(zone).toInstant()

    private fun heartRate(samples: List<Pair<Instant, Long>>) = BleRecordingSampleBuffer(
        heartRateSamples = samples.map { (time, bpm) ->
            BleHeartRateSample(time = time, beatsPerMinute = bpm)
        },
    )

    private fun recorded(bleSamples: BleRecordingSampleBuffer) = ActivityEntryUiState(
        selectedActivityType = DefaultActivityEntryTypes.first(),
        startDateText = "2026-07-14",
        startTimeText = "10:00",
        durationMinutesText = "2",
        isRecordingDraft = true,
        recordedBleSamples = bleSamples,
    )

    @Test
    fun `the end is stretched to cover the last sample`() {
        // The recording ran 10:00:59 -> 10:02:59, rebuilt from text as 10:00 -> 10:02.
        val lastSample = start.plusSeconds(179)
        val request = buildWriteRequest(
            recorded(heartRate(listOf(start.plusSeconds(60) to 120L, lastSample to 150L))),
            ActivityEntryUnits.uniform(UnitSystem.METRIC),
        )

        assertTrue(request!!.endTime.isAfter(lastSample))
        assertEquals(lastSample.plusSeconds(1), request.endTime)
    }

    @Test
    fun `samples before the start are dropped rather than clamped onto it`() {
        // The start is the user's. A sample before it would be clamped onto it, inventing a reading.
        val request = buildWriteRequest(
            recorded(heartRate(listOf(start.minusSeconds(30) to 110L, start.plusSeconds(30) to 120L))),
            ActivityEntryUnits.uniform(UnitSystem.METRIC),
        )

        assertTrue(request!!.bleSamples.isEmpty())
    }

    @Test
    fun `a session already containing its samples is untouched`() {
        val request = buildWriteRequest(
            recorded(heartRate(listOf(start.plusSeconds(30) to 120L, start.plusSeconds(60) to 130L))),
            ActivityEntryUnits.uniform(UnitSystem.METRIC),
        )

        assertEquals(start, request!!.startTime)
        assertEquals(start.plusSeconds(120), request.endTime)
        assertEquals(2, request.bleSamples.heartRateSamples.size)
    }

    @Test
    fun `sample span helpers report the earliest and latest across series`() {
        val buffer = heartRate(
            listOf(start.plusSeconds(10) to 100L, start.plusSeconds(50) to 130L),
        )
        assertEquals(start.plusSeconds(10), buffer.firstSampleTime())
        assertEquals(start.plusSeconds(50), buffer.lastSampleTime())
        assertEquals(null, BleRecordingSampleBuffer().firstSampleTime())
    }
}
